govcd/user.go

```diff
@@ -0,0 +1,463 @@
+)
+
+// Used to check the validity of provider type on creation
+var OrgUserProviderTypes = []string{
+	OrgUserProviderIntegrated,
+	OrgUserProviderSAML,
+	OrgUserProviderOAUTH,
+}
+
+// NewUser creates an empty user
+func NewUser(cli *Client, org *AdminOrg) *OrgUser {
+	return &OrgUser{
+		User:     new(types.User),
+		client:   cli,
+		adminOrg: org,
+	}
+}
+
+// GetUserByNameOrId retrieves an user within an admin organization
+// by either name or ID
+// Returns a valid user if it exists. If it doesn't, returns nil and ErrorEntityNotFound
+func (adminOrg *AdminOrg) GetUserByNameOrId(identifier string, willRefresh bool) (*OrgUser, error) {
+	if willRefresh {
+		err := adminOrg.Refresh()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	for _, orgUser := range adminOrg.AdminOrg.Users.User {
+		if (orgUser.Name == identifier || orgUser.ID == identifier) && orgUser.Type == types.MimeAdminUser {
+
+			user := NewUser(adminOrg.client, adminOrg)
+
+			_, err := adminOrg.client.ExecuteRequest(orgUser.HREF, http.MethodGet,
+				types.MimeAdminUser, "error getting user: %s", nil, user.User)
+
+			return user, err
+		}
+	}
+	return nil, ErrorEntityNotFound
+}
+
+// GetRole finds a role within the organization
+func (adminOrg *AdminOrg) GetRole(roleName string) (*types.Reference, error) {
+
+	// There is no need to refresh the AdminOrg, until we implement CRUD for roles
+	for _, role := range adminOrg.AdminOrg.RoleReferences.RoleReference {
+		if role.Name == roleName {
+			return role, nil
+		}
+	}
+
+	return &types.Reference{}, ErrorEntityNotFound
+}
+
+// CreateUser creates an OrgUser from a full configuration structure
+// The timeOut variable is the maximum time we wait for the user to be ready
+// (This operation does not return a task)
+// Note that the timeout is not absolute, meaning that we only wait the
+// full amount of the timeout if the user has not been created until then.
+// This function returns as soon as the user has been created, which could be as
+// little as 50ms or as much as TimeOut - 50 ms
+// Mandatory fields are: Name, Role, Password.
+// https://code.vmware.com/apis/442/vcloud-director#/doc/doc/operations/POST-CreateUser.html
+func (adminOrg *AdminOrg) CreateUser(userConfiguration *types.User, timeOut time.Duration) (*OrgUser, error) {
+	err := validateUserForCreation(userConfiguration)
+	if err != nil {
+		return nil, err
+	}
+
+	userCreateHREF, err := url.ParseRequestURI(adminOrg.AdminOrg.HREF)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing admin org url: %s", err)
+	}
+	userCreateHREF.Path += "/users"
+
+	user := NewUser(adminOrg.client, adminOrg)
+
+	_, err = adminOrg.client.ExecuteRequest(userCreateHREF.String(), http.MethodPost,
+		types.MimeAdminUser, "error creating user: %s", userConfiguration, user.User)
+	if err != nil {
+		return nil, err
+	}
+
+	// Uncomment this line to show the user structure after creation
+	// ShowUser(*user.OrgUser)
```

Comment:
In general, I'm not keen in leaving such comments for debugging. Either we remove or we add ability to debug at runtime, which would be useful if a user hits something that we need to troubleshoot. What about using some of the debug flags we have already instead?

Reply:
removed